### ISSUE TITLE
feat(Http): implement `toString` for `Headers` type

### ISSUE THREAD

**Issue by gdi2290** [angular/angular#4391]
`.toString()` should serialize into `StringMap<string, string[]>`

update:
this could be done via
```typescript
  toString(): string {
    return this.values().toString();
  }
```
update:
on the client I need to reconstruct server responses 
```typescript
/// <server>
let req = Object.assign({}, res, {
  headers: res.headers.values()
});
/// </server>

 |
 |
index.html
 |
 ▼

/// <client>
let res = Object.assign({}, req, {
  body: req._body,
  headers: new Headers(req.headers)
});
/// </client>
```

**Comment by jeffbcross** [comment]
@gdi2290 does the spec specify this behavior? When I `toString` the native `Headers` object, I get `"[object Headers]"`

**Comment by jeffbcross** [comment]
@gdi2290 btw I'm not fully grasping your use case. Could you break the problem down into a real-world use case?

**Comment by gdi2290** [comment]
#### tldr; 
usecase: On the server during bootstrap we need it serialize each http call on success made on the server. Then we transfer each request to the client by embedding each call into the `index.html` where the client application will replay the calls during the same bootstrap. The replay of the client calls we need to reconstruct each http call as a response and return it immediately.  

toString: Thinking about it more `Headers.toJSON()` makes more sense.

---
usecase:
The use case is to mimic one techniques used in facebook/Relay. The difference is that they do a lot of optimizations within the framework itself to ensure optimum rendering.
> note: React doesn't have control over when the async calls are done. 

What Relay does is pretty clever so I'll try to explain what's happening from what I know. Given this component tree where `r` is our root component.

```
  r      Δ data
 / \     ↓ props
c   c 
   / \   ↓ props
  c   c
```

In react land, the original way of managing data is by sending down the data as props to each component. The problem with that happens when we want to change the data itself and then we end up manually changing the data from the bottom up touching each file. The new "best practice" is for each component to request their own state with the queries declared statically for each component (this is done in facebook/Relay). The technique is interesting because we sorta have this feature already in Angular 1 known as `route resolve` in a `ngRoute` or `state resolve` in `ui-router` if we can imagine only injecting our data by using resolve. 

> I'll try to emulate this technique in Angular 1. For the sake of the explanation let's go with `ui-router` and `state resolve` and I'll be using pseudo code. 

Here's the typical "best practice" module folder structure done in Angular 1 today 

```
root/
  state/
  state/
   state/
   state/
```
Here's an example of a state from ui-router using the "Component" directive design
```es6
.state({
  name: 'home',
  url: '/home',
  controller: HomeCtrl,
  template: '<home></home>',
  resolve: {
    homeData: function($http) {
      return $http.get('/api/home-data');
    }
  }
})
```
>  imagine that `$state` has a magical method called `$state.$routeAsState()`

let's walk the state tree and convert the url to a list of states.
```
   r0.0
   /   \
s1.0   s1.1
      /    \
   s2.0   s2.1
----
/*
route: /r0.0/s0.1/s1.1 
states: [r0.0, s0.1, s1.1]
*/
var routeAsState = $state.$routeAsState(); 
```
>  imagine that we also have another magical method called `$state.$resolve()` that will resolve the state using something like `$q.all`

let's change our root state with a resolve that use our two magic methods
```es6
.state({
  name: 'root',
  url: '/',
  controller: RootCtrl,
  template: '<root></root>',
  resolve: {
    serverData: function($http, $state, $q, $queries) {
      var routeAsState = $state.$routeAsState(); // walk state tree
      var resolvedStates = routeAsState.map(state => state.$resolve($http)) // resolve all data and return as object
      return $q.all(resolvedStates).then($queries.setResolvedData);
    }
  }
})

```
all that's left is injecting `serverData` into our `<root>` directive similar to how we would pass props down a component tree in Angular 2 or react.
```es6
.state({
  name: 'root',
  url: '/',
  controller: RootCtrl,
  template: '<root bind-queries="serverData"></root>',
  resolve: //...
}
function RootCtrl($scope, serverData) {
  $scope.serverData = serverData;
}
```
One last thing, we should replace resolve with another property since we're only resolving in the root. Let's go with queries
```es6
.state({
  name: 'home',
  url: '/home',
  controller: HomeCtrl,
  template: '<home></home>',
  queries: {
    homeData: function($http) {
      return $http.get('/api/home-data');
    }
  }
})
```
we also don't have di here so let's refactor the $http service up to the root and while we're at it let's just remove the function and keep it static metadata.
```es6
function resolveQueries(queries, $q, $http) {
  var dataQueries = {};
  Object.keys(queries).forEach(prop => {
    var method = queries[prop].method;
    var url = queries[prop].url;
    dataQueries[prop] = $http[method](url);
  })
  return $q.all(dataQueries)
}

var resolvedStates = routeAsState.map(state => resolveQueries(state.queries, $q,$http))
```
```es6
.state({
  name: 'home',
  controller: HomeCtrl,
  template: '<home bind-queries="queries"></home>',
  // made up $queries service to inject the data
  resolve: { queries: $queries => $queries.get(this.queries); } } 
  queries: {
    homeData: { method: 'get', url: '/api/home-data' }
  }
})
```
`$queries` is a simple service that will provide the cache version of the data. 

Now that's a lot of boilerplate to implement features that are simpler to do in Angular 2. We end up deleting half the Angular 1 code when writing the same thing in Angular 2.
```typescript
@Component({
  selector: 'home',
  inputs: ['homeData'],
  template: '<div>{{ homeData | async }}</div>'
})
export class Home {
  static queries = {
    'homeData': {
      method: 'get',
      url: '/api/home-data',
      options: {
        headers: new Headers({ auth: 'token' })
      }
    }
  };
  homeData;
  constructor() {
  }
}
```
For facebook/Relay replace the http metadata on each component with GraphQL queries. Here's a simple [Angular 2 + FalcorJS and JSON Graph example](https://gist.github.com/gdi2290/7f9f839504b6b67ebb80). 

so we're assuming the user is not making these optimizations of declaring each ajax call on the component. We can suggest this optimization but we also want to account for the naive case. Solving for the naive case has a lot of constraints which is why we are hijacking `Http` on the server then on success we embed each request to the client as responses inside of the `index.html`.
> `index.html` data transfer with Http server and client is to Http2 as JsonP is to Ajax

One of the benefits of Angular 2 is that it allows for first-class components allowing us access to the view tree as components
 
---
toString:
Thinking about it more `Headers.toJSON()` makes more sense. The problem with returning `"[object Headers]"` is that it's not a native implementation `Headers`. If we had an object that is the native instance of `Headers` and check if the `instanceof` is equal to this implementation then it will be false while `toString` comparison will be the same. The `toString` method in ES5 to compare types to correctly determine different types such as array, object and window.

Here's a crazy thought. Since we use this method of type checking in ES5 we might be able to do the same for something like `Binding` where we might use another di framework. In the past `http` module was removed into it's own module (there was a problem with it during the one or two alpha releases). When it was released there was an angular2 version mismatch where the `ngHttp` module required an older version of angular (in turn a different version of `angular/di` due to how npm2 works). This lead to an error for anyone who tried to use the commonjs version of `ngHttp` since we were presented with a `Binding` error where instance of `Binding` was not correct for `Http` even though it definitely was correct after inserting a break point to check. Since there were two versions of `Binding` (one from the old angular in ngHttp and the other from the correct version of angular) there was an error when using `instanceof` to resolve the bindings as it's two separate modules. If `toString` was used rather than `instanceof` then the next error, if we were using a different di framework, would be a problem with the interface rather than a confusing reference mismatch. We can probably do the same ES5 type check with `Observable` as `obs.toString() === "[object Observable]"` and assume the interface is correct or even make another pass to check that the signature is correct for development version.

Well maybe that was a crazy thought since it probably won't work in Dart and I just want the di module to live in it's own module

**Comment by jeffbcross** [comment]
Recapping a video call I just had with @gdi2290: It's not clear why headers should be serialized from the server to the client for cache. The cache should only contain the original request's method and url (incl params).

I also don't think we should be weakening any instance of checks to do string checks, as we would lose guarantees about the thing being the actual thing we expect. For things like checking if input is an Observable in `async` pipe, it'd be better to check that the object has a `subscribe` method then to check that it's called an Observable. Better yet, in the future it will be better to check that it has a `Symbol.observer` method (or whatever the spec is these days).

Patrick's going to think about the use case some more and get back to me if he thinks of/remembers why this is necessary for cache priming.

**Comment by jeffbcross** [comment]
Ahh, the whole conversation we were talking about `Request` Headers, and that's why I didn't understand why they should be serialized from the server to the client. `Response` Headers are a different story. 

I think `toJson` is reasonable, and we should give thought (outside of this issue) to the right strategy for serializing and processing headers from pre-rendered page to client cache.

**Comment by gdi2290** [comment]
this was merged via https://github.com/angular/angular/pull/4442
